chain/tests/process_block_cut_through.rs

```diff
@@ -0,0 +1,176 @@
+
+	block.header.timestamp = prev.timestamp + Duration::seconds(60);
+	block.header.pow.secondary_scaling = next_header_info.secondary_scaling;
+
+	// If we are skipping roots then just set the header prev_root and skip the other MMR roots.
+	// This allows us to build a header for an "invalid" block by ignoring outputs and kernels.
+	if skip_roots {
+		chain.set_prev_root_only(&mut block.header)?;
+	} else {
+		chain.set_txhashset_roots(&mut block)?;
+	}
+
+	let edge_bits = global::min_edge_bits();
+	block.header.pow.proof.edge_bits = edge_bits;
+	pow::pow_size(
+		&mut block.header,
+		next_header_info.difficulty,
+		global::proofsize(),
+		edge_bits,
+	)
+	.unwrap();
+
+	Ok(block)
+}
+
+#[test]
+fn process_block_cut_through() -> Result<(), chain::Error> {
+	let chain_dir = ".grin.cut_through";
+	global::set_local_chain_type(global::ChainTypes::AutomatedTesting);
+	util::init_test_logger();
+	clean_output_dir(chain_dir);
+
+	let keychain = ExtKeychain::from_random_seed(false)?;
+	let pb = ProofBuilder::new(&keychain);
+	let genesis = genesis_block(&keychain);
+	let chain = init_chain(chain_dir, genesis.clone());
+
+	// Mine a few empty blocks.
+	for _ in 1..6 {
+		let block = build_block(&chain, &keychain, &[], false)?;
+		chain.process_block(block, Options::MINE)?;
+	}
+
+	let key_id1 = ExtKeychainPath::new(1, 1, 0, 0, 0).to_identifier();
+	let key_id2 = ExtKeychainPath::new(1, 2, 0, 0, 0).to_identifier();
+	let key_id3 = ExtKeychainPath::new(1, 3, 0, 0, 0).to_identifier();
+
+	// Build a tx that spends a couple of early coinbase outputs and produces some new outputs.
+	// Note: We reuse key_ids resulting in an input and an output sharing the same commitment.
+	// The input is coinbase and the output is plain.
+	let tx = build::transaction(
+		KernelFeatures::Plain { fee: 0 },
+		&[
+			build::coinbase_input(consensus::REWARD, key_id1.clone()),
+			build::coinbase_input(consensus::REWARD, key_id2.clone()),
+			build::output(60_000_000_000, key_id1.clone()),
+			build::output(50_000_000_000, key_id2.clone()),
+			build::output(10_000_000_000, key_id3.clone()),
+		],
+		&keychain,
+		&pb,
+	)
+	.expect("valid tx");
+
+	// The offending commitment, reused in both an input and an output.
+	let commit = keychain.commit(60_000_000_000, &key_id1, SwitchCommitmentType::Regular)?;
+	let inputs: Vec<_> = tx.inputs().into();
+	assert!(inputs.iter().any(|input| input.commitment() == commit));
+	assert!(tx
+		.outputs()
+		.iter()
+		.any(|output| output.commitment() == commit));
+
+	let verifier_cache = Arc::new(RwLock::new(LruVerifierCache::new()));
+
+	// Transaction is invalid due to cut-through.
+	assert_eq!(
+		tx.validate(Weighting::AsTransaction, verifier_cache.clone()),
+		Err(transaction::Error::CutThrough),
+	);
+
+	// Transaction will not validate against the chain (utxo).
+	assert_eq!(
+		chain.validate_tx(&tx).map_err(|e| e.kind()),
+		Err(chain::ErrorKind::DuplicateCommitment(commit)),
```

Comment:
This appears to be the only test running into a DuplicateCommitment.
It would be good to know the nested functions leading to this in current block processing: i.e.
pipe::process_block
pipe::validate_utxo (has wrong comment /// Validate block inputs against utxo.)
utxo_view::validate_block
utxo_view::validate_output

Reply:
> pipe::validate_utxo (has wrong comment /// Validate block inputs against utxo.)
Fixed comment. Thanks for catching that.